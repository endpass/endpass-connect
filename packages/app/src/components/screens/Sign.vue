<template>
  <SignForm 
    :accounts="accounts"
    :loading="loading"
    :request="request"
    @cancel="handleSignCancel"
    @submit="handleSignSubmit"
  />
</template>

<script>
import { mapActions, mapMutations, mapState } from 'vuex'
import SignForm from '@/components/SignForm.vue';

export default {
  name: 'SignScreen',

  computed: {
    ...mapState({
      accounts: state => state.accounts.accounts,
      inited: state => state.core.inited,
      loading: state => state.core.loading,
      request: state => state.requests.request
    }),
  },

  methods: {
    ...mapActions(['awaitRequestMessage', 'sendMessage', 'closeDialog', 'processRequest']),

    sendMessageAndClose(message) {
      this.sendMessage(message)
      this.closeDialog()
    },

    async handleSignSubmit(res) {
      try {
        await this.processRequest(res.password)
      } catch (err) {
        console.log('sign failed: ', err)
      }
    },

    handleSignCancel() {
      this.sendMessageAndClose({
        message: 'User cancel sign',
        status: false,  
      })
    }
  },

  async created() {
    try {
      // Send ready message to opener for determine application ready
      await this.sendMessage({
        method: 'connect_ready',
        status: true
      })
      this.awaitRequestMessage()
    } catch (err) {
      console.log('Sign created error: ', err)
    }
  },

  components: {
    SignForm
  }
}
</script>