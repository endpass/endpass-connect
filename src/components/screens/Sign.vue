<template>
  <v-frame :loading="!request">    
    <sign-form 
      :accounts="accounts"
      :loading="loading"
      :request="request"
      :error="error"
      :closable="isDialog"
      @cancel="handleSignCancel"
      @submit="handleSignSubmit"
    />
  </v-frame> 
</template>

<script>
import { mapActions, mapGetters, mapState } from 'vuex';
import VFrame from '../VFrame.vue';
import SignForm from '../SignForm.vue';

export default {
  name: 'Sign',

  data: () => ({
    error: null,
  }),

  computed: {
    ...mapState({
      accounts: state => state.accounts.accounts,
      inited: state => state.core.inited,
      loading: state => state.core.loading,
      request: state => state.requests.request,
    }),
    ...mapGetters(['isDialog']),
  },

  methods: {
    ...mapActions(['awaitRequestMessage', 'processRequest', 'cancelRequest']),

    async handleSignSubmit(res) {
      try {
        await this.processRequest(res.password);
        this.error = null;
      } catch (err) {
        this.error = err.message;
      }
    },

    handleSignCancel() {
      this.cancelRequest();
    },

    handleWindowClose() {
      this.cancelRequest();
    },
  },

  async created() {
    if (this.isDialog) {
      window.addEventListener('beforeunload', this.handleWindowClose);
      this.awaitRequestMessage();
    }
  },

  components: {
    SignForm,
    VFrame,
  },
};
</script>
