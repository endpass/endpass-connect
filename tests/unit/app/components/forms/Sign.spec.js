import { shallowMount, mount } from '@vue/test-utils';
import Sign from '@/components/forms/Sign.vue';

const request = {
  address: '0x0',
  url: 'http://localhost',
  request: {
    foo: 'bar',
  },
};

describe('Sign', () => {
  describe('render', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = shallowMount(Sign, {
        propsData: {
          request,
        },
      });
    });

    it('should correctly render Sign component', () => {
      expect(wrapper.name()).toBe('SignForm');
      expect(wrapper.find('[data-test=request-body]').exists()).toBe(true);
      expect(wrapper.find('[data-test=requester-url]').text()).toBe(
        request.url,
      );
      expect(wrapper.find('[data-test=error-message]').exists()).toBe(false);
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should not render requester url if it is not passed', () => {
      wrapper = shallowMount(Sign, {
        propsData: {
          request: {
            ...request,
            url: null,
          },
        },
      });

      expect(wrapper.find('[data-test=requester-url]').exists()).toBe(false);
    });

    it('should not render request body code if it is not passed', () => {
      wrapper = shallowMount(Sign, {
        propsData: {
          request: {
            ...request,
            request: null,
          },
        },
      });

      expect(wrapper.find('[data-test=request-body]').exists()).toBe(false);
    });

    it('should change submit button text if loading and make it disabled', () => {
      wrapper = shallowMount(Sign, {
        propsData: {
          loading: true,
          request,
        },
      });

      const submitButton = wrapper.find('[data-test=submit-button]');

      expect(submitButton.text()).toBe('Loading...');
      expect(submitButton.attributes().disabled).toBe('true');
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('behavior', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = mount(Sign, {
        propsData: {
          request,
        },
      });
    });

    it('should not allow submit form if password is empty', () => {
      wrapper.find('form').trigger('submit');

      expect(wrapper.emitted().submit).toBe(undefined);
    });

    it('should allow submit of email is valid', () => {
      wrapper.setData({
        password: 'foo',
      });

      wrapper.find('form').trigger('submit');

      expect(wrapper.emitted().submit).toEqual([
        [
          {
            password: 'foo',
            account: request.address,
          },
        ],
      ]);
    });

    it('should cancel sign on cancel button press', () => {
      wrapper.find('[data-test=cancel-button]').trigger('click');

      expect(wrapper.emitted().cancel).toEqual([[]]);
    });
  });
});
