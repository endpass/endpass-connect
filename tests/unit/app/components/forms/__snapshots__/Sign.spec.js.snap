// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign render should change submit button text if loading and make it disabled 1`] = `
<form data-test="sign-form">
  <form-field-stub><a href="http://localhost" data-test="requester-url">http://localhost</a>
    requests sign action.
  </form-field-stub>
  <form-field-stub label="Requires request sign by:">
    <message-stub ellipsis="true" data-test="account-address">
      0x0
    </message-stub>
  </form-field-stub>
  <!---->
  <form-field-stub label="Your account password:">
    <v-input-stub placeholder="Enter your password..." name="" value="" autofocus="true" type="password"></v-input-stub>
  </form-field-stub>
  <form-field-stub label="Request data:">
    <v-code-stub data-test="request-body">
      {
      "foo": "bar"
      }
    </v-code-stub>
  </form-field-stub>
  <form-controls-stub>
    <v-button-stub disabled="true" type="primary" submit="true" data-test="submit-button">Loading...</v-button-stub>
    <v-button-stub disabled="true" data-test="cancel-button">Close</v-button-stub>
  </form-controls-stub>
</form>
`;

exports[`Sign render should correctly render Sign component 1`] = `
<form data-test="sign-form">
  <form-field-stub><a href="http://localhost" data-test="requester-url">http://localhost</a>
    requests sign action.
  </form-field-stub>
  <form-field-stub label="Requires request sign by:">
    <message-stub ellipsis="true" data-test="account-address">
      0x0
    </message-stub>
  </form-field-stub>
  <!---->
  <form-field-stub label="Your account password:">
    <v-input-stub placeholder="Enter your password..." name="" value="" autofocus="true" type="password"></v-input-stub>
  </form-field-stub>
  <form-field-stub label="Request data:">
    <v-code-stub data-test="request-body">
      {
      "foo": "bar"
      }
    </v-code-stub>
  </form-field-stub>
  <form-controls-stub>
    <v-button-stub disabled="true" type="primary" submit="true" data-test="submit-button">Sign</v-button-stub>
    <v-button-stub data-test="cancel-button">Close</v-button-stub>
  </form-controls-stub>
</form>
`;
